migrate.py

```diff
@@ -0,0 +1,217 @@
+
+SNAPSHOT_TAG = "snapshot="
+
+DEBUG = False
+
+
+def _run(command: str, input: Optional[str] = None, void_stderr: bool = False) -> None:
+    other_args = dict()
+    if void_stderr and not DEBUG:
+        other_args["stderr"] = subprocess.DEVNULL
+    subprocess.run(command, shell=True, text=True, input=input, **other_args)
+
+
+def _eval(command: str, input: Optional[str] = None, void_stderr: bool = False) -> str:
+    other_args = dict()
+    if void_stderr and not DEBUG:
+        other_args["stderr"] = subprocess.DEVNULL
+    return subprocess.run(command, shell=True, text=True, stdout=subprocess.PIPE, input=input, **other_args).stdout
+
+
+class Backuper:
+
+    def __init__(self,
+                 restic_repo_prefix: str,
+                 zfs_dataset_common_prefix: str,
+                 restic_password_file: str):
+        self.restic_repo_prefix: str = restic_repo_prefix.rstrip("/")
+        self.zfs_dataset_common_prefix: str = zfs_dataset_common_prefix
+        self.restic_password_file: str = restic_password_file
+
+    def _restic_cmd(self, restic_repo: str, restic_command: str, flags: List[str] = []) -> str:
+        initial_args = ["-r", restic_repo, "--password-file", self.restic_password_file, restic_command]
+        args = initial_args + flags
+        arg_string = " ".join([f"'{arg}'" for arg in args])
+        return f"restic {arg_string}"
+
+    def _get_dataset_snapshots(self, dataset_name: str) -> List[str]:
+        lines = _eval(f"sudo zfs list -Hp -o name -t snapshot '{dataset_name}'")
+        snapshots = [line.split("@")[-1] for line in lines.split("\n") if len(line) > 0]
+        return snapshots
+
+    def _get_snapshot_tag(self, datum: Dict[str, Any]) -> str:
+        tags = datum["tags"]
+        tag: str
+        for tag in tags:
+            if tag.startswith(SNAPSHOT_TAG):
+                return tag[len(SNAPSHOT_TAG):]
+        raise Exception("Snapshot does not have a valid snapshot tag.")
+
+    def _get_snapshots_in_restic(self, restic_repo: str) -> Dict[str, str]:
+        json_data = _eval(self._restic_cmd(restic_repo, "snapshots", ["--json"]))
+        data = json.loads(json_data)
+        return {self._get_snapshot_tag(datum): datum["short_id"] for datum in data}
+
+    def _get_zfs_snapshot_size(self, dataset_name: str, snapshot_name: str) -> int:
+        result = _eval(f"sudo zfs list -Hp -o used -t snapshot '{dataset_name}@{snapshot_name}'")
+        return int(result)
+
+    def _get_repo_name_and_path(self, dataset_name) -> Tuple[str, str]:
+        ds_name_without_prefix = dataset_name.removeprefix(self.zfs_dataset_common_prefix).strip("/")
+        repo_name = "/".join([self.restic_repo_prefix, ds_name_without_prefix])
+        path_in_restic_repo = "/" + ds_name_without_prefix
+
+        return repo_name, path_in_restic_repo
+
+    def _init_restic_repo(self, restic_repo):
+        result = _eval(self._restic_cmd(restic_repo, "cat", ["config"]), void_stderr=True)
```

Comment:
I'd prefer the following approach: Just try to `init` the repo.

- Zero exit code -> repo initialized successfully
- Non-zero exit code and "already initialized" in STDERR > repo already exists, also ok
- All other cases -> error, abort!